src/renderer/caching_shaper.rs

```diff
@@ -48,101 +44,150 @@ struct Asset;
 
 const DEFAULT_FONT_SIZE: f32 = 14.0;
 
-#[derive(new, Clone, Hash, PartialEq, Eq, Debug)]
-struct ShapeKey {
-    pub text: String,
-    pub bold: bool,
-    pub italic: bool,
+#[derive(Clone)]
+pub struct ExtendedFontFamily {
+    pub fonts: Vec<SkriboFont>
 }
 
-pub fn add_font_to_collection_by_name(
-    name: &str,
-    source: &SystemSource,
-    collection: &mut FontCollection,
-) -> Option<()> {
-    source
-        .select_family_by_name(name)
-        .ok()
-        .and_then(|matching_fonts| matching_fonts.fonts()[0].load().ok())
-        .map(|font| collection.add_family(FontFamily::new_from_font(font)))
-}
+impl ExtendedFontFamily {
+    pub fn new() -> ExtendedFontFamily {
+        ExtendedFontFamily { fonts: Vec::new() }
+    }
 
-#[cfg(feature = "embed-fonts")]
-pub fn add_asset_font_to_collection(name: &str, collection: &mut FontCollection) -> Option<()> {
-    Asset::get(name)
-        .and_then(|font_data| Font::from_bytes(font_data.to_vec().into(), 0).ok())
-        .map(|font| collection.add_family(FontFamily::new_from_font(font)))
+    pub fn add_font(&mut self, font: SkriboFont) {
+        self.fonts.push(font);
+    }
+
+    pub fn get(&self, props: Properties) -> Option<&Font> {
+        for handle in &self.fonts {
+            let font = &handle.font;
+            let properties = font.properties();
+
+            if properties.weight == props.weight && properties.style == props.style {
+                return Some(&font);
+            }
+        }
+
+        None
+    }
 }
 
-pub fn build_collection_by_font_name(
-    font_name: Option<&str>,
-    bold: bool,
-    italic: bool,
-) -> FontCollection {
-    let source = SystemSource::new();
+fn to_family(x: ExtendedFontFamily) -> FontFamily {
```

Comment:
This can be moved to the above impl block

Reply:
for some reason i did fight a lot with the borrow checker for this but i can't remember why